qiskit/quantum_info/operators/operator.py

```diff
@@ -198,6 +198,99 @@ def from_label(cls, label):
                 op = op.compose(label_mats[char], qargs=[qubit])
         return op
 
+    def apply_permutation(self, perm: list, front: bool = False):
+        """Modifies operator's data by composing it with a permutation.
+
+        Args:
+            perm (list): permutation pattern, describing which qubits
+                occupy the positions 0, 1, 2, etc. after applying the permutation.
+            front (bool): When set to ``True`` the permutation is applied before the
+                operator, when set to ``False`` the permutation is applied after the
+                operator.
+        Returns:
+            Operator: The modified operator.
+
+        Raises:
+            QiskitError: if the size of the permutation pattern does not match the
+                dimensions of the operator.
+        """
+
+        if not front:
+            # The operator is applied first, the permutation is applied after;
+            # however, in terms of matrices, we compute [P][O].
+
+            if len(perm) != len(self._op_shape.dims_l()):
+                raise QiskitError(
+                    "The size of the permutation pattern does not match dimensions of the operator."
+                )
+
+            inv_perm = np.argsort(perm)
+
+            # shape: inv-permuted on left, original on right
+            shape_l = self._op_shape.dims_l()
+            shape_l = shape_l[::-1]
+            shape_l = tuple(shape_l[x] for x in inv_perm)
+            shape_l = shape_l[::-1]
```

Comment:
I've got to say that this _feels_ wrong to me - we seem to be splitting the matrix into dimensions that don't match the actual input.  Could you explain why this should be the case?  Intuitively to me, it feels like the only thing that should vary based on the permutation is the rearrangement of the axes in the middle (the argument to `transpose`).

Assuming you're right, I think a code comment explaining the algorithm would be very helpful here.

Just as a minor legibility not - lines 230 to 233 might be clearer as:
```python
raw_shape_l = self._op_shape.dims_l()
n_dims_l = len(raw_shape_l)
shape_l = tuple(raw_shape_l[n_dims_l - n - 1] for n in reversed(inv_perm))
```
rather than reversing the tuple twice.  The whole thing still feels mathematically odd to me, as described above, though.

Reply:
I have applied these and the following suggestions, making the code simpler and more readable, thanks!